Add weight examples and semantics

ajreynol opened the pull request.

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -1428,10 +1428,10 @@ \subsection{Setting Benchmark Information}
 A benchmark whose header contains the line
 \[
-\paren{\setinfokwd\mbox{ }:{\tt sygus\mbox{-}version\mbox{ }2.0}}
+\paren{\setinfokwd\mbox{ }:{\tt sygus\mbox{-}version\mbox{ }2.1}}
```
Should the 2.0 on line 1433 below also be 2.1?

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
+let $G$ be a grammar provided for function-to-synthesize $f$, and let $W$ be a weight keyword.
+Let $y$ be a non-terminal symbol of $G$ and let $t$ be a term appearing
+in the grouped rule listing for $y$, as described in the syntax for grammar
+definitions in Section~\ref{ssec:dec-synth-fun}.
```
To be consistent with other references, please use `\cref{ssec:dec-synth-fun}` here.

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
 \label{sec:weight-semantics}
 For each function-to-synthesize $f$ and weight keyword $W$
 as mentioned in Section~\ref{sec:weight-attributes},
```
To be consistent with other references, please use `\cref{ssec:weight-attributes}` here.

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
+where $y_i \mapsto_{W,k_i} t_i$ is a rule in $G$.
+
+Notice that since there may be \emph{multiple} sequences of terms that
+generates the same term $r$ above.
```
s/generates/generate

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -1781,18 +1781,55 @@ \subsubsection{Weights}
+replacing an occurrence of some $y_i$ by $t_i$
+where $y_i \mapsto_{W,k_i} t_i$ is a rule in $G$.
+
+Notice that since there may be \emph{multiple} sequences of terms that
```
May be ~~since~~?

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -2476,6 +2513,63 @@ \section{Examples}%
+\begin{lstlisting}[language=SyGuS]
+(set-logic NIA)
+(set-feature :weights true)
+(declare-weight numX)
```
Please also add `declare-weight` to `keywords=[6]` set in `listing-style.tex` for syntax highlighting

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -2476,6 +2513,63 @@ \section{Examples}%
+(constraint (and (> (f x) x) (<= (_ numI f) 1)))
+(check-synth)
+\end{lstlisting}
+In this example, a weight keyword $\texttt{numRec}$ has been declared,
```
Did you mean `numI` in place of `numRec` here?

SaswatPadhi commented on language/sygus.tex:
```diff
@@ -1058,7 +1058,7 @@ \subsection{Declaring Sorts}
 \end{itemize}
 
-\subsection{Declaring Weight}
+\subsection{Declaring Weight Keywords}
```
Please also add `declare-weight` to the reserved words in appendix A.

SaswatPadhi reviewed (changes requested):
Thanks Andy.

I have a few comments, all minor though:

ajreynol commented:
I've addressed the comments, the reserved words I've also updated to include the oracle commands.  Also, I've improved the 2nd example to be a bit more intuitive.

SaswatPadhi commented:
Thanks! Looks great. Yay! Now we have support for weighted SyGuS :)

Merging this.
